### PR TITLE
Poimandres Draft

### PR DESCRIPTION
**Pull request by Sajad-Sharhani**
@dai-shi first draft, what do you think?
Any suggestions for use cases?

### CODE REVIEW

**Comment by dai-shi** [comment]
As I understand, we want a single doc for all three zustand/jotai/valtio, right?

**Comment by Sajad-Sharhani** [comment]
the thing that I understood is that markerikson will make a category for `pmndrs` that will include a page for every lib.
right?

**Comment by Sajad-Sharhani** [comment]
I think for the first step, it's more logical to introduce them separately, and after that, we will become more flexible to do a variety of combinations

**Comment by dai-shi** [comment]
I think he mean a single page, but you can confirm him if you aren't sure.

**Review by dai-shi — COMMENTED** [review]
I think it's a good start. It's likely that people would like to learn the real "use cases" of the library, which is not well answered.

**Comment by Sajad-Sharhani** [comment]
Do you have any specific keywords for the use cases in your mind?

**Comment by dai-shi** [comment]
The question people would ask is "When should I use zustand?" Would be nice if we answer to it.

**Comment by Sajad-Sharhani** [comment]
@dai-shi , what do you think?

File: website/docs/state-management/zustand.mdx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+## Further Information
+
+- [📚 Micro State Management with React Hooks](https://www.amazon.com/Micro-State-Management-React-Hooks-dp-1801812373/dp/1801812373/ref=mt_other?_encoding=UTF8&me=&qid=1643728846)
+- [Zustand GitHub repository](https://github.com/pmndrs/zustand)
+- [Zustand demo](https://zustand-demo.pmnd.rs/)
+- [Jotai GitHub repository](https://github.com/pmndrs/jotai)
+- [📚 Jotai documentation](https://jotai.org/docs/introduction)
+- [Valtio demo](https://github.com/pmndrs/valtio)
+- [📚 Valtio documentation](https://valtio-demo.pmnd.rs/)
```
Comment (dai-shi) [inline website/docs/state-management/zustand.mdx]: Do you mean this?
```suggestion
- [Valtio GitHub repository](https://github.com/pmndrs/valtio)
- [Valtio demo](https://valtio-demo.pmnd.rs/)
```

**Review by dai-shi — COMMENTED** [review]
Overall it looks good. I will read carefully again later for correctness.
Meanwhile, I would like someone to review this too mainly for readability. I'll see if someone can help.
Also, what do you think adding small ToDo apps for comparison?

**Comment by Sajad-Sharhani** [comment]
I think we can add examples in the next step. I don't think it fits right now.
to be consistence to how they present the pages

**Comment by dai-shi** [comment]
> I'll see if someone can help.

I asked @DennisSmolek for help.

**Comment by DennisSmolek** [comment]
What is the status of the project in general? Just getting acquainted with the page/project and want to make sure all your hard work doesn't go somewhere stale.

**Comment by dai-shi** [comment]
You mean the project https://github.com/markerikson/react-community-tools-practices-cheatsheet?

It's not certain, but they try to launch. We don't know the future, but if there were better places, we could migrate. At this point, I think it's worth moving forward.

**Comment by DennisSmolek** [comment]
@dai-shi I'll go over the page in details some time later today hopefully but as it's yet another library comparison I feel some of the best text on this comes from your book. 
Are you ok with me quoting passages in Open Source resources like this?
Or should I re-word things?

**Comment by dai-shi** [comment]
@DennisSmolek I think we can quote it under fair use. (Note that I don't have copyright.)

**Comment by dai-shi** [comment]
@Sajad-Sharhani So sorry for the delay.
Seems like @DennisSmolek is busy.

Let me mention @dsacramone who could be helpful (unless I mistake the username).

Otherwise, I'll see if someone else can.

File: website/docs/state-management/poimandres.mdx [hunk 2]
```diff
@@ -0,0 +1,124 @@
+---
+id: poimandres
+title: "Poimandres"
+sidebar_label: "Poimandres"
+---
+
```
Comment (dai-shi) [inline website/docs/state-management/poimandres.mdx:6]: Maybe we should add some notes about Poimandres.
```suggestion

[Poimandres](https://github.com/pmndrs) is a developer collective and they host various OSS projects. Some of famous projects are [react-spring](https://github.com/pmndrs/react-spring) and [react-three-fiber](https://github.com/pmndrs/react-three-fiber). They also provide three popular state managers, namely [zustand](https://zustand-demo.pmnd.rs), [jotai](https://jotai.org), and [valtio](https://valtio.pmnd.rs). This document introduces those three state managers and discuss their features.

```

File: website/docs/state-management/poimandres.mdx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+# Zustand
+
+## Description
+
+Zustand is a tiny library primarily designed to create module state for React. It's based on an immutable update model, in which state objects can't be modified but always have to be newly created. Render optimization is done manually using selectors. It has a straightforward and yet powerful store creator interface.
+
+## Purpose and Use Cases
+
+Zustand is a library that's used to create a store that holds a state. It's primarily designed for module state, which means you define this store in a module and export it. It's based on the immutable state model, in which you are not allowed to modify state object properties. Updating states must be done by creating new objects, while unmodified state objects must be reused. The benefit of the immutable state model is that you only need to check state object referential equality to know if there's any update; you don't have to check equality deeply.
```
Comment (dai-shi) [inline website/docs/state-management/poimandres.mdx]: This seems a bit repetitive with the previous Description section. How about moving some description in the previous section, and this "Purpose and Use Cases" section to discuss more details.

Some points to cover:
- Zustand is very small. So, it's good for a lightweight solution. For example, some apps use react-query for server state, and such apps only need a small global state. Zustand would be a best fit for such small use cases.
- Immutable state model is the same as what React itself is based on. There would be less mental model switch between React state and zustand global state. So, it would be good for developers who are used to React's useState.
- Unopinionated. The library has less restrictions, and very open. It's developers responsibility to organize well. It would be best if developers need flexibility with responsibility.

Something like that. There should be some others. Feel free to have some more.

File: website/docs/state-management/poimandres.mdx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+## Tradeoffs
```
Comment (dai-shi) [inline website/docs/state-management/poimandres.mdx:24]: This is tradeoffs among z/j/v, right? People might expect tradeoffs from some other libs, like redux. So, it would be nice to make it clear.

File: website/docs/state-management/poimandres.mdx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+Zustand – or any other libraries with this approach – is a simple addition to the React principle. It's a good recommendation if you need a library with a small bundle size, if you are familiar with referential equality and memoization, or you prefer manual render optimization.
+
+# Jotai
+
+## Description
+
+Jotai is a small library for the global state. It's modeled after useState/useReducer and with what are called atoms, which are usually small pieces of state. Unlike Zustand, it is a component state, and like Zustand, it is an immutable update model.
+
+## Purpose and Use Cases
```
Comment (dai-shi) [inline website/docs/state-management/poimandres.mdx:78]: Hmm, maybe the section title doesn't match the content?

File: website/docs/state-management/poimandres.mdx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+const countAtom = atom(0);
+```
+
+Notice the similarity with `useState(0)`. `useAtom(countAtom)` returns the same tuple, `[count, setCount]`,
+as `useState(0)` does.
+
+### Dynamic atom creation
+
+The second benefit of Jotai is a new capability—that is, dynamic atom creation. Atoms can be created and destroyed in the React component lifecycle. This is not possible with the multiple-Context approach, because adding a new state means adding a new `Provider` component. If you add a new component, all its child components will be remounted, throwing away their states.
+
```
Comment (dai-shi) [inline website/docs/state-management/poimandres.mdx:112]: Actually, the third benefit would be "derived atoms".

File: website/docs/state-management/poimandres.mdx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+An `atom` represents a piece of a state. An `atom` is usually a small piece of state, and it is a minimum unit of triggering re-renders. The `atom` function creates a definition of an atom. The `atom` function takes one argument to specify an initial value, just as `useState` does. The following code is used to define a new atom:
+
+```js
+const countAtom = atom(0);
+```
+
+Notice the similarity with `useState(0)`. `useAtom(countAtom)` returns the same tuple, `[count, setCount]`,
+as `useState(0)` does.
+
+### Dynamic atom creation
```
Comment (dai-shi) [inline website/docs/state-management/poimandres.mdx]: I would rephrase it like "Non-global state", or something.

File: website/docs/state-management/poimandres.mdx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+The second benefit of Jotai is a new capability—that is, dynamic atom creation. Atoms can be created and destroyed in the React component lifecycle. This is not possible with the multiple-Context approach, because adding a new state means adding a new `Provider` component. If you add a new component, all its child components will be remounted, throwing away their states.
+
+## When Should I Consider Using This?
+
+My answer whould be to use jotai for component-centric apps.
```
Comment (dai-shi) [inline website/docs/state-management/poimandres.mdx]: Shouldn't it be "my" or "our". Let's make it look like facts rather than opinions.

File: website/docs/state-management/poimandres.mdx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+## When Should I Consider Using This?
+
+My answer whould be to use valtio for data-centric apps.
+
+The data-centric approach is you have data first regardless of React components. React components are used to represent those data. For example, in game development, it’s likely that you may have game state in advance to design components. You don’t want these data to be controlled by React lifecycle.
+
+# Comparing Zustand, Jotai, and Valtio
+
+First of all, all three libraries are provided by the Poimandres GitHub organization. It's a developer collective providing many libraries. Three micro-state management libraries from a single GitHub organization may sound counter-intuitive, but they are in different styles. There is also a philosophy that is common in the three libraries: their small API surfaces. All three libraries try their best to provide small API surfaces and let developers compose the APIs as they want.
```
Comment (dai-shi) [inline website/docs/state-management/poimandres.mdx]: 👍

**Review by dai-shi — COMMENTED** [review]
Thanks for working on this!
My concern is how readers will read this doc. They would be interested in "Use Cases" which is weak at the moment. There would be some other aspects wanted from readers. 🤔

File: website/docs/state-management/poimandres.mdx (hunk 4 above)
Comment (Sajad-Sharhani) [reply website/docs/state-management/poimandres.mdx:24]: I don't see the confusion.
the tradeoff is for it as a state management lib as a whole.

**Review by dai-shi — COMMENTED** [review]
It can be the next step and a different PR, but I would definitely like to see code examples (plus codesandbox links) of the three with the same app to compare.

**Comment by sandren** [comment]
Thank you for this! This question comes up a lot and a doc like this could help resolve a lot of frequently asked questions. This is a great start.🙂

I think adding a brief overview at the top would be helpful. We could explain why an open source collective would provide three different state management libraries instead of one (to provide alternatives for all three major paradigms: flux, proxy, and atomic models) and how each relates to the wider React state management ecosystem (how they are inspired by Redux, MobX, and Recoil respectively).

Also I agree with @dai-shi that examples of the same application written with each of the three different libraries would go a long way in demonstrating the differences. [This CodeSandbox](https://codesandbox.io/s/zustand-valtio-jotai-counters-my6np?file=/src/App.tsx) of his could be used as reference.

**Comment by dai-shi** [comment]
Thanks @sandren !

> We could explain why an open source collective would provide three different state management libraries instead of one...

Probably, I would need to write something?

**Comment by Sajad-Sharhani** [comment]
> I think we can add examples in the next step. I don't think it fits right now.
> to be consistence to how they present the pages

But if u think  example is needed in the current PR I will do it.

https://react-community-tools-practices-cheatsheet.netlify.app/state-management/redux

**Comment by dai-shi** [comment]
@Sajad-Sharhani 
I do think we need code snippets to compare 3 lib styles (they don't have to be playable codesandbox examples), before submitting a PR to the official repo.

**Review by dai-shi — APPROVED** [review]
Thanks for the adjustment. It's been a few months since we started this. Let's move forward. Please open a PR to the main repo and get some feedbacks.
